test/e2e/tests/errors.spec.js

```diff
@@ -1,8 +1,70 @@
+const { resolve } = require('path');
+const { promises: fs } = require('fs');
 const { strict: assert } = require('assert');
+const { get, has, set } = require('lodash');
 const { Browser } = require('selenium-webdriver');
+const { format } = require('prettier');
 const { convertToHexValue, withFixtures } = require('../helpers');
 const FixtureBuilder = require('../fixture-builder');
 
+const dateFields = ['metamask.conversionDate'];
+
+/**
+ * Transform date properties to value types, to ensure that state is
+ * consistent between test runs.
+ *
+ * @param {unknown} data - The data to transform
+ */
+function transformDates(data) {
+  for (const field of dateFields) {
+    if (has(data, field)) {
+      set(data, field, typeof get(data, field));
+    }
+  }
+  return data;
+}
+
+/**
+ * Check that the data provided matches the snapshot.
+ *
+ * @param {object }args - Function arguments.
+ * @param {any} args.data - The data to compare with the snapshot.
+ * @param {string} args.snapshot - The name of the snapshot.
+ * @param {boolean} [args.update] - Whether to update the snapshot if it doesn't match.
+ */
+async function matchesSnapshot({
+  data: unprocessedData,
+  snapshot,
+  update = process.env.UPDATE_SNAPSHOTS === 'true',
+}) {
+  const data = transformDates(unprocessedData);
+
+  const snapshotPath = resolve(__dirname, `./state-snapshots/${snapshot}.json`);
+  const rawSnapshotData = await fs.readFile(snapshotPath, {
+    encoding: 'utf-8',
+  });
+  const snapshotData = JSON.parse(rawSnapshotData);
+
+  try {
+    assert.deepStrictEqual(data, snapshotData);
+  } catch (error) {
+    if (update && error instanceof assert.AssertionError) {
+      const stringifiedData = JSON.stringify(data);
+      // filepath specified so that Prettier can infer which parser to use
+      // from the file extension
+      const formattedData = format(stringifiedData, {
+        filepath: 'something.json',
+      });
+      await fs.writeFile(snapshotPath, formattedData, {
+        encoding: 'utf-8',
+      });
+      console.log(`Snapshot '${snapshot}' updated`);
+      return;
+    }
+    throw error;
+  }
```

Comment:
I was a bit confused by this function, wondering why we don't communicate anything about the `deepStrictEqual` failure and why we just update the snapshot instead of failing the test. But then I re-read the PR description and the default value of the `update` param and this makes sense. We do throw there error except in cases where `UPDATE_SNAPSHOTS` has been explicitly set, which would be done intentionally when the snapsshots are to be updated.

:+1: